smart_open/gcs.py

```diff
@@ -0,0 +1,523 @@
+            # twice here, once in the if condition and once when calling index.
+            #
+            # This is sub-optimal, but better than the alternative: wrapping
+            # .index in a try..except, because that is slower.
+            #
+            remaining_buffer = self._buffer.peek()
+            if self._line_terminator in remaining_buffer:
+                next_newline = remaining_buffer.index(self._line_terminator)
+                the_line.write(self._read_from_buffer(next_newline + 1))
+                break
+            else:
+                the_line.write(self._read_from_buffer())
+                self._fill_buffer()
+        return the_line.getvalue()
+
+    def terminate(self):
+        """Do nothing."""
+        pass
+
+    #
+    # Internal methods.
+    #
+    def _read_from_buffer(self, size=-1):
+        """Remove at most size bytes from our buffer and return them."""
+        # logger.debug('reading %r bytes from %r byte-long buffer', size, len(self._buffer))
+        size = size if size >= 0 else len(self._buffer)
+        part = self._buffer.read(size)
+        self._current_pos += len(part)
+        # logger.debug('part: %r', part)
+        return part
+
+    def _fill_buffer(self, size=-1):
+        size = size if size >= 0 else self._buffer._chunk_size
+        while len(self._buffer) < size and not self._eof:
+            bytes_read = self._buffer.fill(self._raw_reader)
+            if bytes_read == 0:
+                logger.debug('reached EOF while filling buffer')
+                self._eof = True
+
+    def __str__(self):
+        return "(%s, %r, %r)" % \
+              (self.__class__.__name__, self._bucket.name, self._blob.name)
+
+    def __repr__(self):
+        return (
+            "%s("
+            "bucket=%r, "
+            "blob=%r, "
+            "buffer_size=%r)"
+        ) % (
+            self.__class__.__name__,
+            self._bucket.name,
+            self._blob.name,
+            self._buffer_size,
+        )
+
+
+class BufferedOutputBase(io.BufferedIOBase):
+    """Writes bytes to GCS.
+
+    Implements the io.BufferedIOBase interface of the standard library."""
+
+    def __init__(
+            self,
+            bucket,
+            blob,
+            min_part_size=_DEFAULT_MIN_PART_SIZE,
+            client=None,  # type: google.cloud.storage.Client
+    ):
+        if client is None:
+            client = google.cloud.storage.Client()
+        self._client = client
+        self._credentials = self._client._credentials
+        self._bucket = self._client.bucket(bucket)  # type: google.cloud.storage.Bucket
+        self._blob = self._bucket.blob(blob)  # type: google.cloud.storage.Blob
+        assert min_part_size % _MIN_MIN_PART_SIZE == 0, 'min part size must be a multiple of 256KB'
+        self._min_part_size = min_part_size
+
+        self._total_size = 0
+        self._total_parts = 0
+        self._buf = io.BytesIO()
```

Comment:
I think a better name would be:

```suggestion
        self._current_part = io.BytesIO()
```